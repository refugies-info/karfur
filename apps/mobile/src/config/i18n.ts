import { Languages } from "@refugies-info/api-types";

export const fallback: Languages = "fr";
export const supportedLocales: Record<
  Languages,
  { translationFileLoader: Function }
> = {
  fr: {
    translationFileLoader: () => require("../translations/fr.json"),
  },
  en: {
    translationFileLoader: () => require("../translations/en.json"),
  },
  ar: {
    translationFileLoader: () => require("../translations/ar.json"),
  },
  ti: {
    translationFileLoader: () => require("../translations/ti.json"),
  },
  ru: {
    translationFileLoader: () => require("../translations/ru.json"),
  },
  ps: {
    translationFileLoader: () => require("../translations/ps.json"),
  },
  fa: {
    translationFileLoader: () => require("../translations/fa.json"),
  },
  uk: {
    translationFileLoader: () => require("../translations/uk.json"),
  },
};
