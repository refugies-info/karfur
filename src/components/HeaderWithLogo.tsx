import React from "react";
import { View } from "react-native";
import { Icon } from "react-native-eva-icons";
import { useSafeAreaInsets } from "react-native-safe-area-context";
import { useRoute } from "@react-navigation/native";
import styled from "styled-components/native";

import { styles } from "../theme";
import Logo from "../theme/images/logo.svg";

import { logEventInFirebase } from "../utils/logEvent";
import { FirebaseEvent } from "../utils/eventsUsedInFirebase";

import { useTranslationWithRTL } from "../hooks/useTranslationWithRTL";

import { SmallButton } from "./SmallButton";
import { RowContainer, RTLView } from "./BasicComponents";
import { LanguageSwitch } from "./Language/LanguageSwitch";
import NotificationsIcon from "./Notifications/NotificationsIcon";
import { StyledTextSmallBold } from "./StyledText";

const MainContainer = styled(RowContainer)`
  padding-horizontal: ${styles.margin * 3}px;
  align-items: center;
  justify-content: space-between;
  display: flex;
  padding-top: ${styles.margin}px;
`;
const StyledText = styled(StyledTextSmallBold)`
  margin-left: ${(props: { isRTL: boolean }) =>
    props.isRTL ? 0 : styles.margin}px;
  margin-right: ${(props: { isRTL: boolean }) =>
    props.isRTL ? styles.margin : 0}px;
`;

const LOGO_WIDTH = 58;
const LOGO_HEIGHT = 40;
const ICON_SIZE = 24;

interface Props {
  onLongPressSwitchLanguage?: () => void;
  hideLanguageSwitch?: boolean;
  hideLogo?: boolean;
  text?: string;
  iconName?: string;
}
export const HeaderWithLogo = ({
  onLongPressSwitchLanguage,
  hideLanguageSwitch,
  hideLogo,
  text,
  iconName,
}: Props) => {
  const { isRTL } = useTranslationWithRTL();
  const insets = useSafeAreaInsets();
  const { name: routeName } = useRoute();

  return (
    <View style={{ paddingTop: insets.top, paddingBottom: styles.margin }}>
      <MainContainer isRTL={false}>
        {!hideLogo ? (
          <Logo
            width={LOGO_WIDTH}
            height={LOGO_HEIGHT}
            accessible={true}
            accessibilityLabel="Réfugiés point info"
          />
        ) : (
          <View />
        )}
        {text && (
          <RTLView
            style={{
              flex: 1,
              marginRight: hideLanguageSwitch && !isRTL ? styles.margin * 7 : 0,
              marginLeft: hideLanguageSwitch && isRTL ? styles.margin * 7 : 0,
              justifyContent: "center",
            }}
          >
            {iconName && (
              <Icon
                name={iconName}
                width={ICON_SIZE}
                height={ICON_SIZE}
                fill={styles.colors.black}
              />
            )}
            <StyledText isRTL={isRTL}>{text}</StyledText>
          </RTLView>
        )}
        <RowContainer>
          {routeName === "ExplorerScreen" && <NotificationsIcon />}
          {!hideLanguageSwitch && (
            <LanguageSwitch
              onLongPressSwitchLanguage={() => {
                logEventInFirebase(
                  FirebaseEvent.LONG_PRESS_CHANGE_LANGUAGE,
                  {}
                );
                if (onLongPressSwitchLanguage) {
                  onLongPressSwitchLanguage();
                }
              }}
            />
          )}
        </RowContainer>
      </MainContainer>
    </View>
  );
};

interface PropsBack {
  onLongPressSwitchLanguage?: () => void;
  navigation: any;
  backScreen?: string;
}
export const HeaderWithBackForWrapper = ({
  onLongPressSwitchLanguage,
  navigation,
  backScreen,
}: PropsBack) => {
  const insets = useSafeAreaInsets();
  const { t } = useTranslationWithRTL();

  return (
    <View style={{ paddingTop: insets.top }}>
      <MainContainer isRTL={false}>
        <SmallButton
          iconName="arrow-back-outline"
          onPress={
            !!backScreen
              ? () => {
                  navigation.popToTop();
                  navigation.navigate(backScreen);
                }
              : navigation.goBack
          }
          label={t("global.back_button_accessibility")}
        />
        <RowContainer>
          <LanguageSwitch
            onLongPressSwitchLanguage={() => {
              logEventInFirebase(FirebaseEvent.LONG_PRESS_CHANGE_LANGUAGE, {});
              if (onLongPressSwitchLanguage) {
                onLongPressSwitchLanguage();
              }
            }}
          />
        </RowContainer>
      </MainContainer>
    </View>
  );
};
