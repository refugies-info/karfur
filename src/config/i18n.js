export const fallback = "fr";
export const supportedLocales = {
  fr: {
    translationFileLoader: () => require("../translations/fr.json"),
  },
  en: {
    translationFileLoader: () => require("../translations/en.json"),
  },
  ar: {
    translationFileLoader: () => require("../translations/ar.json"),
  },
  "ti-ER": {
    translationFileLoader: () => require("../translations/ti-ER.json"),
  },
  ru: {
    translationFileLoader: () => require("../translations/ru.json"),
  },
  ps: {
    translationFileLoader: () => require("../translations/ps.json"),
  },
  fa: {
    translationFileLoader: () => require("../translations/fa.json"),
  },
};
