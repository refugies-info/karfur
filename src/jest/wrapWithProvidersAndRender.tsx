import React from "react";
import { render } from "react-native-testing-library";
import { Provider } from "react-redux";
import { createStore } from "redux";
import {
  initialRootStateFactory,
  rootReducer,
  RootState,
} from "../services/redux/reducers";

interface WrapWithProvidersAndRenderParams {
  Component: React.FunctionComponent<unknown>;
  compProps?: Record<string, unknown>;
  reduxState?: Partial<RootState>;
}

/**
 * Provide all the ugly-to-set-up providers for your component to be ready to test
 * @param Component
 * @param compProps properties passed on to [getPropsWithNavigation()] to gen navigation props
 * @param reduxState defaults to initialRootState
 */
export function wrapWithProvidersAndRender({
  Component,
  compProps = {},
  reduxState = initialRootStateFactory(),
}: WrapWithProvidersAndRenderParams) {
  const store = createStore(rootReducer, reduxState);

  const componentWithReduxAndI18n = (
    <Provider store={store}>
      <Component {...compProps} />
    </Provider>
  );

  return render(componentWithReduxAndI18n);
}
