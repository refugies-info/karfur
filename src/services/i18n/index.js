import i18next from "i18next";
import * as config from "../../config/i18n";
import translationLoader from "./translation-loader";
const i18n = {
  /**
   * @returns {Promise}
   */
  init: () => {
    return new Promise((resolve, reject) => {
      i18next.use(translationLoader).init(
        {
          fallbackLng: config.fallback,

          interpolation: {
            escapeValue: false,
          },
        },
        (error) => {
          if (error) {
            return reject(error);
          }
          resolve();
        }
      );
    });
  },
  use: (elem) => i18next.use(elem),
  /**
   * @param {string} key
   * @param {Object} options
   * @returns {string}
   */
  t: (key, options) => i18next.t(key, options),
  /**
   * @returns {string}
   */
  get locale() {
    return i18next.language;
  },
  /**
   * @returns {'LTR' | 'RTL'}
   */
  get dir() {
    return i18next.dir().toUpperCase();
  },
  /**
   * @returns {boolean}
   */
  isRTL: () => {
    return i18next.language ? i18next.dir().toUpperCase() === "RTL" : false;
  },
  /**
   * Similar to React Native's Platform.select(),
   * i18n.select() takes a map with two keys, 'rtl'
   * and 'ltr'. It then returns the value referenced
   * by either of the keys, given the current
   * locale's direction.
   *
   * @param {Object<string,mixed>} map
   * @returns {mixed}
   */
  select(map) {
    const key = this.isRTL ? "rtl" : "ltr";
    return map[key];
  },

  changeLanguage: (ln) => i18next.changeLanguage(ln),
};
export const t = i18n.t;
export default i18n;
