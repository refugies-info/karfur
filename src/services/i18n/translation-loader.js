import * as config from "../../config/i18n";
const translationLoader = {
  type: "backend",
  init: () => {},
  read: function (language, _, callback) {
    let resource,
      error = null;
    try {
      resource = config.supportedLocales[language].translationFileLoader();
    } catch (_error) {
      error = _error;
    }
    callback(error, resource);
  },
};
export default translationLoader;
