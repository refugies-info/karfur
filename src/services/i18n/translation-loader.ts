import { Languages } from "@refugies-info/api-types";
import * as config from "../../config/i18n";

const translationLoader: any = {
  type: "backend",
  init: () => {},
  read: function (language: Languages, _: any, callback: Function) {
    let resource,
      error = null;
    try {
      resource = config.supportedLocales[language].translationFileLoader();
    } catch (_error) {
      error = _error;
    }
    callback(error, resource);
  },
};
export default translationLoader;
