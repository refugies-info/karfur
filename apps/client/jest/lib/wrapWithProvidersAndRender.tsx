import { TooltipProvider } from "@radix-ui/react-tooltip";
import { render } from "@testing-library/react";
import { initialMockStore } from "__fixtures__/reduxStore";
import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { RootState } from "services/rootReducer";

interface WrapWithProvidersAndRenderParams<Props> {
  Component: React.FunctionComponent<any>;
  compProps?: Props;
  reduxState?: Partial<RootState>;
}

/**
 * Provide all the ugly-to-set-up providers for your component to be ready to test
 * @param Component
 * @param compProps properties passed on to [getPropsWithNavigation()] to gen navigation props
 * @param reduxState defaults to initialRootState
 */

export function wrapWithProvidersAndRenderForTesting<T>({
  Component,
  compProps,
  reduxState = initialMockStore,
}: WrapWithProvidersAndRenderParams<T>) {
  const mockStore = configureStore([]);
  const store = mockStore(reduxState);

  const componentWithRedux = (
    <TooltipProvider>
      <Router>
        <Provider store={store}>
          <Component {...compProps} />
        </Provider>
      </Router>
    </TooltipProvider>
  );

  return render(componentWithRedux);
}
