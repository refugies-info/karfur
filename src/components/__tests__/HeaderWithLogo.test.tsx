import { wrapWithProvidersAndRender } from "../../jest/wrapWithProvidersAndRender";
import { HeaderWithLogo } from "../HeaderWithLogo";
import i18n from "../../services/i18n";

jest.mock("../../services/i18n", () => ({
  __esModule: true, // this property makes it work
  default: { isRTL: jest.fn() },
}));

describe("HeaderWithLogo", () => {
  it("should render correctly with LTR", () => {
    i18n.isRTL.mockReturnValueOnce(false);
    const component = wrapWithProvidersAndRender({ Component: HeaderWithLogo });
    expect(component).toMatchSnapshot();
  });

  it("should render correctly with RTL", () => {
    i18n.isRTL.mockReturnValueOnce(true);
    const component = wrapWithProvidersAndRender({ Component: HeaderWithLogo });
    expect(component).toMatchSnapshot();
  });
});
