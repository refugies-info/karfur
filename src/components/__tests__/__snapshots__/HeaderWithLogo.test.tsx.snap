// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderWithLogo should render correctly with LTR 1`] = `
<View
  style={
    Object {
      "paddingBottom": 8,
      "paddingTop": undefined,
    }
  }
>
  <View
    isRTL={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 24,
          "paddingTop": 8,
        },
      ]
    }
  >
    <react-native-svg
      accessibilityLabel="Réfugiés point info"
      height={40}
      width={58}
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
          },
        ]
      }
    >
      <View />
    </View>
  </View>
</View>
`;

exports[`HeaderWithLogo should render correctly with RTL 1`] = `
<View
  style={
    Object {
      "paddingBottom": 8,
      "paddingTop": undefined,
    }
  }
>
  <View
    isRTL={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 24,
          "paddingTop": 8,
        },
      ]
    }
  >
    <react-native-svg
      accessibilityLabel="Réfugiés point info"
      height={40}
      width={58}
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
          },
        ]
      }
    >
      <View />
    </View>
  </View>
</View>
`;
